James/deployments click to deploy tutorial (#12478)
pulumi/docs: content/tutorials/pulumi-deployments-click-to-deploy/index.md

---

> title_tag: Create and Manage cloud infrastructure with Pulumi Deployments | Learn Pulumi

Reviewer: Tag and title dont match

Author: Fixed.

title: "Create Infrastructure with Pulumi Deployments Click-to-Deploy"
layout: single
description: |
    Create infrastructure deployments using the New Project Wizard and GitHub integration to enable push-to-deploy in the Pulumi Cloud.
meta_desc: In this step-by-step tutorial, learn how to create and manage cloud infrastructure using Pulumi Deployments, the New Project Wizard, and GitHub integration.
meta_image: meta.png
weight: 50
summary: |

>     In this tutorial, you'll learn how to create, deploy, and manage cloud infrastructure using Pulumi Deployments with Click-to-Deploy. You'll start by using the New Project Wizard to set up and launch infrastructure from your browser, selecting a template to get started quickly. Additionally, you'll configure the Pulumi GitHub App to support additional deployment triggers, including push-to-deploy.

Reviewer: How different is this than the current conceptual docs? Is this net new? Are we deprecating the old thing?

Author: Net-new. We don't have anything this deep; just a short paragraph here: https://www.pulumi.com/docs/pulumi-cloud/deployments/reference/#click-to-deploy. Would not be depreciating anything old.

Author: That's the plan, once complete I'll be redirecting so it's not two pages.

you'll_learn:
  - How to use the New Project Wizard to create and deploy a Pulumi project from your browser
  - How to install and configure the Pulumi GitHub App
  - How to trigger and manage deployments directly from the Pulumi Cloud console using click-to-deploy
prereqs:
    - Completion of the [Getting Started](/docs/get-started/) guide or familiarity with the basics of the Pulumi workflow.
    - The [Pulumi CLI](/docs/install/)
    - A [Pulumi Cloud account](https://app.pulumi.com/signup)

>     - An [Amazon Web Services](https://aws.amazon.com/) account

Reviewer 2: Probably worth mentioning you'll need programmatic access and credentials in some form as well. (See my other comment about setting environment variables.)

    - A [GitHub account](https://github.com/) with admin rights to a Git repository or organization
    - Familiarity with JavaScript, TypeScript, or Python
estimated_time: 10
allow_long_title: true
---

> ## What is Pulumi Deployments

Author: fixed

[Pulumi Deployments](/docs/pulumi-cloud/deployments/) is a fully managed service designed to automate cloud infrastructure. Deployments provide the building blocks designed for scaling your cloud infrastructure, with tools to power public and private cloud platforms, APIs, and self-service infrastructure portals. These options provide flexibility to incorporate Pulumi Deployments into your existing workflows.

> Deployments can be triggered through multiple methods: using the [REST API](/docs/pulumi-cloud/deployments/reference/#deployments-rest-api), integrating with GitHub via [push-to-deploy](/docs/pulumi-cloud/deployments/reference/#github-app-installation), or initiating directly from the [Pulumi Cloud Console](/docs/pulumi-cloud/deployments/reference/#pulumi-console).

Author: fixed

> ## Setup a new Pulumi Deployment

Author: fixed

First sign in to Pulumi Cloud via the [Pulumi sign in page](https://app.pulumi.com) or by running `pulumi login` via the Pulumi CLI.

## Install and configure the GitHub App

Next, install and configure the [Pulumi GitHub App](/docs/using-pulumi/continuous-delivery/github-app/) to enable Pulumi Deployment's New Project Wizard and push-to-deploy functionality. Follow the GitHub app [installation instructions](/docs/pulumi-cloud/deployments/reference/#github-app-installation) to ensure your Pulumi organization is set up to work seamlessly with Pulumi Deployments.

## Create a New Project with the New Project Wizard

> To create a new Pulumi project, commit its code, and deploy it entirely from your browser, you will use the [New Project Wizard](/docs/pulumi-cloud/developer-portals/new-project-wizard/). The wizard allows you to start from one of Pulumi’s public templates, your [Organization Templates](/docs/pulumi-cloud/developer-portals/templates), or even generate a custom project using the cloud resources and your preferred language of choice with [Pulumi AI](link).

Reviewer 2: We've mentioned the New Project Wizard several times at this point, but this is the first time we're linking to it. We should link to it earlier on.

> First, navigate to the **New Project** tab in the Pulumi Cloud console. Choose your source template from Pulumi, then you can optionally choose your cloud and language to filter the results or search all templates.

Reviewer 2: We should make this a numbered list rather than a succession of headings, as it's a series of steps to be taken in order. (I know the current docs present it this way, but IMO that's not the right way to present it.)

![Animation of how to create a new project in Pulumi Cloud](./pulumi-create-new-project.gif)

{{% notes type="info" %}}
If your Pulumi administrator has configured [custom templates](/docs/pulumi-cloud/developer-portals/templates) for your organization, you will be able to choose from your organization's custom templates from your own image source.
If you select "Use a template" but your organization doesn't have custom templates, you can choose one of Pulumi's public templates to get started.
{{% /notes %}}

Type `aws-python` or `aws-typescript` and click **next**.

### Project details

> Provide a name for the project, along with an optional description and the stack name which will be set in the resulting `Pulumi.yaml` file and the name of `Pulumi.<stack>.yaml`.

Reviewer 2: ```suggestion
Provide a name for the project, an optional description, and a name for the project's first stack. The project name will be written to the project file (`Pulumi.yaml`) and the stack name will be used to name the stack-configuration file (`Pulumi.<stack>.yaml`).
```

### Configuration

> Provide any values required for the configuration declared in the [template](/docs/pulumi-cloud/developer-portals/templates), in this example, you can set the **aws:region** you would like to deploy into.

Author: thx!

These values populate the `config` section in the `Pulumi.<stack>.yaml` file, specifying the necessary settings for your project.

### Environment

> If you've configured [environments](/docs/pulumi-cloud/esc) with your organization, you can specify one to use with the resulting stack. Environments enable teams to manage hierarchical collections of configuration and secrets without needing to re-specify all of them during project creation.

Reviewer 2: The word "hierarchical" feels superfluous here, given the tutorial doesn't use any hierarchical config. 

```suggestion
If you've configured [environments](/docs/pulumi-cloud/esc) with your organization, you can specify one to use with the resulting stack. Environments enable teams to define and share collections of configuration across multiple Pulumi projects and stacks.
```

{{% notes type="info" %}}

> Secret environment variables, such as `AWS_SECRET_ACCESS_KEY`, are encrypted end-to-end and can be set on each stack. However, by creating an environment with Pulumi ESC, you can centralize secrets and set up OIDC for secure authentication. This allows you to manage and share sensitive configuration data across multiple stacks efficiently.

Reviewer 2: This note seems out of place. It has to do with setting setting environment variables, but since the section is about choosing an ESC environment (which it assumes you've already configured), it doesn't really apply. (Or it's not clear to me how it applies.)

{{% /notes %}}

### Repository

Configure the repository and optional subdirectory to use when committing your new project code.

{{% notes "info" %}}
If you granted the Pulumi GitHub app access to _all_ repositories, the New Project Wizard will allow users to create projects with Deployments enabled in new repositories.

If the app only has access to _some_ repositories, users will only be able to create new projects with Deployments enabled in _existing_ repositories.
{{% /notes %}}

> The GitHub owner is not configurable, as that must match the Pulumi GitHub app's owner in order to work with Deployments.

Reviewer 2: Feels like this could be clearer. Might be worth mentioning that if the GitHub organization you want to write to isn't the one listed, you'll need to use a different Pulumi organization. (As I believe the Pulumi orgs and GitHub app installations are 1:1.)

Author: Made a new suggestion, take a look!

### Deployment settings

Configure the behavior of Deployments, including when to trigger new Deployments and any environment variables required for the Pulumi program to run.

> A full description of each setting is available [here](/docs/pulumi-cloud/deployments/reference/#deployment-settings).

Reviewer 2: Technically fine, but it's better for both SEO and usability to link using relevant text  rather than "click [here](#)".

Author: thx

### Deployment method

Finally, select Pulumi Deployments as the deployment method and **Create project** to trigger a deployment. You will see your new deployment start within your **Stacks Deployment** tab.

![Animation of how to create a new project in Pulumi](./pulumi-new-project-wizard.gif)

> When your deployment is complete, you can now manage and deploy your infrastructure from the Pulumi Cloud Console. In your stack's **Actions** tab, use the **[Click to Deploy](https://www.pulumi.com/docs/pulumi-cloud/deployments/reference/#click-to-deploy)** button to trigger a deployment.

Reviewer 2: I'm honestly not sure we should assume the happy path here. I don't know about you, but this almost never works for me on the first try; there's just about always something else that has to be done to get to a green deployment. 

Indeed, as currently written, this tutorial _can't_ work unless the user pastes their AWS credentials into the UI as environment variables at setup time. Some users will be reluctant to do this, but there's really no way for this work without it. We should directly address rather than sidestep this IMO.

This feature is useful for day-to-day operational tasks, such as debugging a stuck stack or rectifying drift by performing a refresh. With Click to Deploy, you can execute these actions without needing to pull the stack and its source code onto your local machine.

![Image of the Pulumi cloud console actions](./pulumi-cloud-actions.png)

## Cleanup

To avoid incurring any unwanted charges, clean up the resources you created:

1. Navigate to your stack in the Pulumi console.
2. Delete the stack, by clicking on the **Settings** tab and the **Delete stack** button.
3. Delete the GitHub repository you created.

## Next steps

> In this tutorial, you set up a basic push-to-deploy stack using Pulumi’s new project wizard. You configured your stack to use the GitHub app, deployed resources, and explored the fundamental Pulumi Deployment browser workflow.

Reviewer 2: A natural next step would also be to push a change to your newly provisioned GitHub repo and see the deployment get triggered! And/or submit a pull request and see a PR preview.

Author: Yup totally and that's the next tutorial which i'll be linking to.

---
title_tag: Create Infrastructure with Pulumi Deployments Click-to-Deploy | Learn Pulumi
title: "Create Infrastructure with Pulumi Deployments Click-to-Deploy"
layout: single
description: |
    Create and deploy infrastructure in the browser with Pulumi Deployments and the New Project Wizard.
meta_desc: In this step-by-step tutorial, learn how to create and manage cloud infrastructure using Pulumi Deployments, the New Project Wizard, and GitHub integration.
meta_image: meta.png
weight: 50
summary: |

>     In this tutorial, you'll learn how to create, deploy, and manage cloud infrastructure using Pulumi Deployments with click-to-deploy. You'll start by using the New Project Wizard to set up and launch infrastructure from your browser, selecting a template to get started quickly. Additionally, you'll configure the Pulumi GitHub App to support additional deployment triggers, including push-to-deploy.

Reviewer 2: I thought I called this out before, but we don't actually "configure the Pulumi GitHub App to support additional deployment triggers, including push-to-deploy". All we do is have them install the GitHub app. (Sorry to be nickpicky on this, but it always bugs me when docs like this suggest they'll cover something and then don't.)

Author: thx

youll_learn:
  - How to use the New Project Wizard to create and deploy a Pulumi project from your browser
  - How to install and configure the Pulumi GitHub App

>   - How to trigger and manage deployments directly from the Pulumi Cloud console using click-to-deploy

Reviewer 2: ```suggestion
  - How to trigger deployments directly from the Pulumi Cloud console using click-to-deploy
```

Author: thx

prereqs:
    - Completion of the AWS [Getting Started](/docs/clouds/aws/get-started/) guide or familiarity with the basics of the Pulumi workflow
    - A [Pulumi Cloud account](https://app.pulumi.com/signup)

>     - AWS credentials configured

Reviewer 2: My first question on reading this would be "configured how"? Local configuration won't matter of course, since we won't be running locally. We should explain what "configured" means or just remove it, and leave it at "AWS credentials", or "An AWS acccount, access key ID, and secret access key". (The last version also mentioned needing an AWS account, which we seem to have removed.)

    - Create a [Pulumi Cloud account](https://app.pulumi.com/signup)
    - An [Amazon Web Services](https://aws.amazon.com/) account
    - A [GitHub account](https://github.com/) with admin rights to a git repository or organization
    - You have familiarity with JavaScript, TypeScript, or Python
estimated_time: 10
collections:

>     - deployments-fundamentals

Author: thx; skipping the collection for now

If you select "Use a template" but your organization doesn't have custom templates, you'll be able to choose one of Pulumi's public templates.
{{% /notes %}}

### Project details

Provide a name for the project, along with an optional description and the stack name which will be set in the resulting `Pulumi.yaml` file and the name of `Pulumi.<stack>.yaml`.

### Configuration

Provide any values required for the configuration declared in the [template](/docs/pulumi-cloud/developer-portals/templates), in this example, you can set the`aws:region` you would like to deploy into.

These values populate the `config` section in the `Pulumi.<stack>.yaml` file, specifying the necessary settings for your project.

### Environment

If you've configured [environments](/docs/pulumi-cloud/esc) with your organization, you can specify one to use with the resulting stack. Environments enable teams to manage hierarchical collections of configuration and secrets without needing to re-specify all of them during project creation.

> {{% notes "info" %}}

Author: done

### Environment

If you've configured [environments](/docs/pulumi-cloud/esc) with your organization, you can specify one to use with the resulting stack. Environments enable teams to define and share collections of configuration and secrets across multiple Pulumi projects and stacks, such as your `AWS_SECRET_ACCESS_KEY` which will be needed to configure your deployment settings. 

> ### Repository

Reviewer 2: Just a note to ask about provisioning private repos (as folks are likely to wonder this) and/or what to do if we only support creating public repos.

Author: all add this to my docs feedback notes; then come back and clarify when we update them.

![Image of the Pulumi cloud console actions](./pulumi-cloud-actions.png)

> ### Destroy all stack resources

Reviewer 2: This should be a sibling of Create a Project, so an H2:

```suggestion
## Destroy all stack resources
```

After you are finished exploring your new stack, you can begin to clean up the resources created. In the **Actions** tab, select the **Destroy** option to remove all resources associated with the stack. Confirm by clicking the **Delete resources?** confirmation box, then click **Deploy**. Pulumi will begin tearing down the resources, and you can track the progress directly in the console.

> If you no longer need the stack itself, you can also delete it. In the same **Settings** tab, choose **Delete stack** to remove the stack, including all its deployment history and state files from Pulumi Cloud.

Reviewer 2: ```suggestion
If you no longer need the stack itself, you can also delete it. In the **Settings** tab, choose **Delete stack** to remove the stack, including all its deployment history and state files from Pulumi Cloud.
```

Author: thx, fixed

> ## Cleanup

Reviewer 2: There's no cost for having public repos on GitHub, so I'd probably just delete this section or pull the bit about deleting the repo into the section above this one, and maybe call that "Clean Up" instead (which is where we usually put destroy steps like these). (Which reminds me -- is it possible to create a new project with a _private_ repo? 🤔 We should probably call out that repos are public by default, or perhaps they're in line with your organization's default repo-creation settings?)

Author: In my testing its not something you can set in the pulumi UI, it defaults to my settings, in this case I happened to create a private org, so they've all been private.
